pkg/controller/pxc/users_custom.go

```diff
@@ -0,0 +1,365 @@
+		}
+	}
+
+	for _, db := range desired.DBs {
+		if !current.DBs.Has(db) {
+			log.Info("DBs changed", "current", current.DBs, "desired", desired.DBs, "user", userName)
+			return true
+		}
+	}
+
+	for _, host := range desired.Hosts {
+		if _, ok := current.Grants[host]; !ok && len(desired.Grants) > 0 {
+			log.Info("Grants for user host not present", "host", host, "user", userName)
+			return true
+		}
+
+		for _, grant := range desired.Grants {
+			for _, currGrant := range current.Grants[host] {
+				if currGrant == fmt.Sprintf("GRANT USAGE ON *.* TO `%s`@`%s`", desired.Name, host) {
+					continue
+				}
+
+				if !strings.Contains(currGrant, strings.ToUpper(grant)) {
+					log.Info("Grant not present in current grants", "grant", grant, "user", userName)
+					return true
+				}
+
+				if desired.WithGrantOption && !strings.Contains(currGrant, "WITH GRANT OPTION") {
+					log.Info("Grant with grant option not present", "user", userName)
+					return true
+				}
+			}
+		}
+
+		for _, db := range desired.DBs {
+			dbPresent := false
+
+			for _, currGrant := range current.Grants[host] {
+				if strings.Contains(currGrant, fmt.Sprintf("ON `%s`.*", db)) {
+					dbPresent = true
+					break
+				}
+			}
+
+			if !dbPresent {
+				log.Info("DB not present in current grants", "db", db, "user", userName)
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// getUserSecret gets secret by name defined by `user.PasswordSecretRef.Name` or returns a secret
+// with newly generated password if name matches defaultName
+func getUserSecret(ctx context.Context, cl client.Client, cr *api.PerconaXtraDBCluster, name, defaultName, passKey string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	err := cl.Get(ctx, types.NamespacedName{Name: name, Namespace: cr.Namespace}, secret)
+
+	if err != nil && name != defaultName {
+		return nil, errors.Wrap(err, "failed to get user secret")
+	}
+
+	if err != nil && !k8serrors.IsNotFound(err) && name == defaultName {
+		return nil, errors.Wrap(err, "failed to get default user secret")
+	}
+
+	if err != nil && k8serrors.IsNotFound(err) {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: cr.Namespace,
+			},
+		}
+		err := generateUserPass(ctx, cl, cr, secret, passKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate user password secrets")
+		}
+
+		return secret, nil
+	}
+
+	_, hasPass := secret.Data[passKey]
+	if !hasPass && name == defaultName {
+		pass, err := generatePass()
+		if err != nil {
+			return nil, errors.Wrap(err, "generate custom user password")
+		}
+
+		if secret.Data == nil {
+			secret.Data = make(map[string][]byte)
+		}
+
+		secret.Data[passKey] = pass
+
+		err = cl.Update(ctx, secret)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to update user secret")
+		}
+	}
+
+	// pass key should be present in the user provided secret
+	if !hasPass {
+		return nil, errors.New("password key not found in secret")
+	}
```

Comment:
This will basically happen only if `!hasPass && name != defaultName`, which means we will not create anything at this moment, just check if it is missing from the provided secret.

Reply:
This will also happen if `!hasPass && name == defaultName`. The first if statement doesn't return after the update. This means that we will go into the second if statement and return an error.

Reply:
Ok ok, I see.